Reorg package, update for interfaces, and cleanup docs

econchick opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines <https://github.com/spotify/gordon-janitor-gcp/blob/master/CONTRIBUTING.rst>
2. If the PR is unfinished, please prefix the subject line with [WIP], [DRAFT], or [RFC].
-->

**What this PR does / why we need it**:

So sorry this is so many commits - hopefully it isn't cognitively too much.

I've reorganized the repo/package structure as well as renamed a few classes as previously discussed in person. I've also flattened the namespacing of the package so that rendered HTML docs don't have such loooong paths to the documented object. I reorg'ed the factory function for the publisher client, and added one for the reconciler. I split up `gcp_clients` module to follow convention of a client per module. And then a bunch of touch ups to the docs.

**Which issue(s) this PR fixes**
<!-- optional; in `fixes #<issue number>, fixes #<issue_number>, ...` format, will close the issue(s) when PR gets merged: -->

N/A

**Special notes for your reviewer**:

Sorry @matt0bi  😬
 
**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, start the release note with the string "action required: ".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

econchick committed:
Reorg repo structure

econchick committed:
Add aiohttp intersphinx mapping link

So whenever we mention an `aiohttp.ClientSession` is required, it automatically links to aiohttp’s docs \o/

econchick committed:
Flatten namespace for readable docs

econchick committed:
Add small touchups for docs

econchick committed:
Reorg docs layout to mimic package layout

econchick committed:
Rename classes for conciseness

econchick committed:
Split gcp client module into gce and gcrm

In order to follow convention of 1 client per module.

econchick committed:
Move get_pub factory function into plugin’s init

Anticipating all public factory functions to be in plugin/__init__

econchick committed:
Add reconciler factory function

econchick committed:
Add MOAR doc touchups

I can’t shake this perfectionism.

matt0bi commented on docs/plugins.rst:
```diff
@@ -0,0 +1,31 @@
+.. attention::
+
+    These plugins are internal modules for the core `gordon-janitor`_
+    logic. No other use cases are expected.
```
:grey_question: Do we have a unified rST style guide? I'm asking because some lines in the docs look like they follow the Python 80-char rule and some don't, and Sphinx doesn't care about it because afaik it dumps the whole into a `<p></p>` block.

econchick replied:
Yeah I unnecessarily added a line break here, sure. 

The 80-char rule is actually ambiguous with the google style guide - it doesn't state the max length. pep8 does limit to 72 chars, which I have visual lines on my editor for and just instinctively follow.

matt0bi commented on gordon_janitor_gcp/clients/auth.py:
```diff
@@ -72,9 +76,9 @@ class GoogleAuthClient:
-            NOTE: if ``session`` is given, then ``loop`` will be ignored.
-            Otherwise, ``loop`` will be used to create a session, if
-            provided.
+            NOTE: if :obj:`session` is given, then ``loop`` will be
```
If I understand rst and the intent here, then the \`\`loop\`\` word on this line should also have been turned into ":obj:\`loop\`" (?)

matt0bi commented on gordon_janitor_gcp/clients/auth.py:
```diff
@@ -137,9 +141,9 @@ def _setup_token_request(self):
 
         Raises:
-            exceptions.GCPAuthError: if no token was found in the
+            :exc:`.GCPAuthError`: if no token was found in the
```
TIL about [sphinx domains](http://www.sphinx-doc.org/en/master/domains.html) :+1:

matt0bi commented on gordon_janitor_gcp/clients/gce.py:
```diff
@@ -48,83 +47,31 @@
+            session object attached to :obj:`auth_client` if not
+            provided.
         api_version (str): version of API endpoint to send requests to.
-        blacklisted_tags (list): Do not collect an instance if it has been
```
:grey_question: Did `been` and `metadata` need to be moved to another line? I'm asking because I'm trying to figure out the pattern to use moving forward.

econchick replied:
Talked about in person - I follow pep8's docstring/comment line length (72) since it's not explicitly defined in Google's style guide.

matt0bi commented on gordon_janitor_gcp/plugins/__init__.py:
```diff
@@ -0,0 +1,76 @@
+        A :class:`GDNSReconciler` instance.
+    """
+    reconciler._validate_dns_config(config)
+    auth_client = reconciler._init_dns_auth(config)
```
Should `reconciler._init_dns_auth(config)` function still use an underscore to indicate it's private if it's being called from a different module? (Same question for the other `_init_*` funcs)

econchick replied:
Also talked about this in person - I didn't want these functions to be used or otherwise public.

But I also agreed to make a proper Builder class (in the reconciler module) for `get_reconciler` to interact with, similar to [@matto's approach](https://github.com/spotify/gordon-janitor-gcp/pull/12/files#diff-5f1882cb69773e6c69eb948668bd3502R56).

matt0bi commented on tests/unit/clients/test_gdns.py:
```diff
@@ -50,16 +50,16 @@ def test_create_gcp_rrset():
 
 def test_dns_client_default(mocker):
-    auth_client = mocker.Mock(auth.GoogleAuthClient, autospec=True)
+    auth_client = mocker.Mock(auth.GAuthClient, autospec=True)
```
I know this wasn't added in this PR, but I found out that `mocker.Mock` doesn't support the [autospec](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock) kwarg. Just passing the class as the first arg makes mocker spec out the class - the `autospec` kwarg just adds an attribute to the mock itself: `auth_client.autospec == True`, which can be confusing for other folks.

Wrote some quick code to verify:

```python
import pytest


class Thing:
    def __init__(self, a):
        self.a = a

    def work(self):
        print('processing {}'.format(self.a))


def test_a(mocker):
    mocked_thing = mocker.Mock(Thing, autospec=True)

    autospecced_methods = dir(mocked_thing)
    # dir(mocked_thing)
    [
        '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__',
        '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__',
        '__init__', '__init_subclass__', '__le__', '__lt__', '__module__',
        '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__',
        '__setattr__', '__sizeof__', '__str__', '__subclasshook__',
        '__weakref__', 'assert_any_call', 'assert_called',
        'assert_called_once', 'assert_called_once_with', 'assert_called_with',
        'assert_has_calls', 'assert_not_called', 'attach_mock', 'autospec',
        'call_args', 'call_args_list', 'call_count', 'called',
        'configure_mock', 'method_calls', 'mock_add_spec', 'mock_calls',
        'reset_mock', 'return_value', 'side_effect', 'work'
    ]

    mocked_thing_2 = mocker.Mock(Thing)

    non_autospecced_methods = dir(mocked_thing_2)
    # dir(mocked_thing_2)
    [
        '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__',
        '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__',
        '__init__', '__init_subclass__', '__le__', '__lt__', '__module__',
        '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__',
        '__setattr__', '__sizeof__', '__str__', '__subclasshook__',
        '__weakref__', 'assert_any_call', 'assert_called',
        'assert_called_once', 'assert_called_once_with', 'assert_called_with',
        'assert_has_calls', 'assert_not_called', 'attach_mock', 'call_args',
        'call_args_list', 'call_count', 'called', 'configure_mock',
        'method_calls', 'mock_add_spec', 'mock_calls', 'reset_mock',
        'return_value', 'side_effect', 'work'
    ]
    # mocked_thing has 'autospec=True' attribute, mocked_thing_2 doesn't

    print(set(autospecced_methods) - set(non_autospecced_methods))
    # outputs '{"autospec"}'
    assert len(autospecced_methods) == len(non_autospecced_methods)
    # fails at 47 == 46
    assert autospecced_methods == non_autospecced_methods
```

econchick replied:
lol TIL, thank you.

matt0bi commented on README.rst:
```diff
@@ -16,7 +16,7 @@ Requirements
 
 * Python 3.6. Support for other Python versions may be added in the future.
-* Service account JSON key that has Cloud DNS read access. See Google's `documentation`_ on how to create a key.
+* Service account JSON key that has read access to the plugin service you're wanting to use (e.g. Google Cloud DNS, Pub/Sub, or Compute Engine). See Google's `documentation`_ on how to create a key.
```
> Service account JSON key that has read access to the plugin service

I think the level of access can differ between plugins - like the pub/sub plugin needs write access to post messages?

econchick replied:
d'oh - yeah that was an oversight.

matt0bi commented on gordon_janitor_gcp/plugins/__init__.py:
```diff
@@ -0,0 +1,76 @@
+)
+
+
+def get_publisher(config, changes_channel, **kw):
```
Could you explain why you chose to put these functions in this module vs. keeping them with the classes they use (eg. publisher.py)?

econchick replied:
talked about this in person - I added this here mostly to make the entrypoint in `setup.py` is shorter.

matt0bi commented on gordon_janitor_gcp/plugins/reconciler.py:
```diff
@@ -57,28 +58,61 @@
+
+def _init_dns_client(auth_client, config):
+    kwargs = {
+        'project': config['project'],
```
Could this be a potential KeyError?

econchick replied:
Yep, forgot to add that to the validation of config. Once added there, there shouldn't be a problem here.

matt0bi reviewed (commented):
Had a handful of questions and pointers. I think this change unifies the naming nicely :tada:.

This didn't fit anywhere, so adding it here:
Do you think the two doc's related commit messages (5d2f127 and 7445ea2) could be changed to follow the "{verb} {subject}" (?) pattern we've been using eg. "Touch up docs" instead of "Small touchups for docs"?

econchick committed:
Remove unnecessary linebreaks in prose docs

econchick committed:
Remove unnecessary 'autospec' kwarg in mocks

econchick committed:
Create builder classes for publisher and reconciler

econchick commented:
Thanks @matt0bi ! I think I've addressed everything - let me know if I haven't. I've also added three new commits (sorry!):

* removed the `autospec` keyword in the mocks' kwargs
* removed the unnecessary linebreaks in the rst docs
* created builder classes (no need for new tests since no changes to the public API and still 100% coverage, \o/)

matt0bi commented on gordon_janitor_gcp/plugins/publisher.py:
```diff
@@ -73,64 +73,76 @@
-    if not config.get('topic').startswith(topic_prefix):
-        config['topic'] = f'{topic_prefix}{config["topic"]}'
+class GPubsubPublisherBuilder:
+    """TODO"""
```
Is this docstring a "todo in the future" or a left-over todo?

econchick replied:
Dangit good catch.

matt0bi commented on gordon_janitor_gcp/plugins/reconciler.py:
```diff
@@ -69,33 +69,46 @@
-        msg = 'The GCP project where Cloud DNS is located is required.'
-        logging.error(msg)
-        raise exceptions.GCPConfigError(msg)
+class GDNSReconcilerBuilder:
```
Should this class also have a docstring?

econchick replied:
Indeed it should, thank ya :D

matt0bi reviewed (commented):
The changes look good, I just found some places in the last commit that could use a docstring. 

Also:
> (no need for new tests since no changes to the public API and still 100% coverage, \o/)

Awesome :tada:

econchick commented:
@matt0bi I addressed your two comments - mind a quick re-review?
